### PR TITLE
fixes #244 - Added CountryTweetMap

### LINKED ISSUE #244: Add CountryTweetMap app
## Features of the app
* The app will plot distribution of tweets related to a given query on a map using different colored
  markers to differentiate between high, medium and low number of tweets. These categories will
  be added as layers on the map and the users can switch between them.

* The app will also show frequency of tweets for individual country in tabular format and also visualize it for better understanding.

* Users will be able to apply filters like start and end date and tweet count.

### PR DESCRIPTION
**Pull request by djmgit**
fixes issue #244. Added CountryTweetApp.
User can enter query word and search. Tweet distribution
related to entered query will be plotted on a map
and categorised according to frequency.

I have:
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only strictly only one commit per issue.

### For the reviewers
I have:
- [ ] Reviewed this pull request by an authorized contributor.
- [ ] The reviewer is assigned to the pull request.

### CODE REVIEW

**Comment by djmgit** [comment]
Test link: http://loklakapps.surge.sh/CountryTweetMap/

@kavithaenair @singhpratyush @Achint08 @SKrPl open for review

File: CountryTweetMap/scripts/app.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        var countryHighGroup = L.layerGroup(countriesHigh);
+        var countryMediumGroup = L.layerGroup(countriesMedium);
+        var countryLowGroup = L.layerGroup(countriesLow);
+
+        var background = L.tileLayer(
+            'http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
+            {maxZoom: 18});
+        $scope.map = L.map('map', {
+            center: [mapCenterLat, mapCenterLon],
+            zoom: 1,
```
Comment (kavithaenair) [inline CountryTweetMap/scripts/app.js]: Change the value here, so that there is only 1 map, instead of 3. 3 maps looks weird.

Comment (djmgit) [inline CountryTweetMap/scripts/app.js]: @kavithaenair Sure, I will change it.

**Comment by djmgit** [comment]
@kavithaenair Presently I have not implemented error detection. Also country v/s tweet frequency visualization is also left to be done. This is the basic version of the app. I have thought of iterating over this.

@vibhcool @singhpratyush Loklak is presently returning incorrect lat/long values. As @kavithaenair mentioned, in this app if we search ```kavitha``` then India is being plotted, but its being plotted on the wrong place.
http://api.loklak.org/api/search.json?q=kavitha In the result returned by this api, it can be seen that the co-ordinated of India are incorrect.

**Comment by djmgit** [comment]
@kavithaenair I have fixed the repeated map problem, also I have applied max bounds.

**Comment by singhpratyush** [comment]
@djmgit: We'll need to take a closer look into it. Can you please open an issue in https://github.com/loklak/loklak_server and explain the details there so that we could follow up in a more planned manner?

**Comment by djmgit** [comment]
@singhpratyush sure :)

**Comment by djmgit** [comment]
@singhpratyush I have created issue https://github.com/loklak/loklak_server/issues/1310 for this problem.
Please review the rest of the PR :)

**Comment by singhpratyush** [comment]
@djmgit: I get the following error after 3-4 searches - 

```
angular.min.js:108 Error: Map container is being reused by another instance
    at e.remove (leaflet.js:5)
    at n.$scope.reset (app.js:173)
    at app.js:46
    at angular.min.js:120
    at n.$eval (angular.min.js:134)
    at n.$digest (angular.min.js:132)
    at n.$apply (angular.min.js:135)
    at l (angular.min.js:87)
    at F (angular.min.js:91)
    at angular.min.js:92
```

**Comment by djmgit** [comment]
@singhpratyush I am unable to reproduce the problem, I have tested it on both Firefox and Chrome

**Comment by djmgit** [comment]
@kavithaenair I have added initial map

Test link: http://loklakapps.surge.sh/CountryTweetMap/

**Comment by kavithaenair** [comment]
@djmgit , as @singhpratyush said i get the same error. As you can see in the console in the screenshot.
image

**Comment by djmgit** [comment]
Test link: http://loklakapps.surge.sh/CountryTweetMap/
@singhpratyush @kavithaenair I have made changes, please re-review and check whether the error is still there or not :)

**Comment by kavithaenair** [comment]
@djmgit , after searching for a query, to see the markers i have to scroll over the map.
image

**Comment by djmgit** [comment]
@kavithaenair I have reduced the default zoom level and I have also kept option for further zoom out. Almost all the markers are visible now. I am also thinking of increasing the vertical map space by making some changes to the app layout so that the map can take full width also. however the design is not yet finalized. I will make further changes to the layout in the next iteration as this PR has become quite long and is sitting here for quite a long time

**Comment by djmgit** [comment]
@singhpratyush I was thinking of removing the navbar,  and include a drawable sidebar and put the nav options there. That should give some more space to the map.   What do you think about it?

**Comment by singhpratyush** [comment]
@djmgit: That is also worth a shot!

**Comment by djmgit** [comment]
@singhpratyush ohk,  i will change the layout in the next iteration,  lets see what serves best.

**Comment by vibhcool** [comment]
@djmgit , there is a need to scroll. 
screenshot from 2017-07-17 16-21-18

Browser: Firefox 54.0

File: CountryTweetMap/scripts/app.js [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        var rangeMid = ($scope.tweetMax + $scope.tweetMin) / 2;
+
+        var countriesHigh = [];
+        var countriesMedium = [];
+        var countriesLow = [];
+        for (var key in $scope.tweetFreq) {
+            var country = $scope.tweetFreq[key][0];
+            var size = $scope.tweetFreq[key].length;
+            if (size === $scope.tweetMax) {
+                var marker = L.marker(country.place_country_center, {icon: redIcon}).bindPopup(
```
Comment (vibhcool) [inline CountryTweetMap/scripts/app.js]: @djmgit , please see this snippet, if this looks better. 😅 
```
if (size > rangeMid) {
var not_a_marker = L.marker(country.place_country_center, {icon: blueIcon});
}
else if (size <=rangeMid) {
//something
} else if (size === rangeMid) {
//something else
}
var marker = not_a_marker.bindPopup(
                    "<div class='header'>" + country.place_country +
                    " (" + country.place_country_code + ")" + "</div><div class='tweet-number'>" +
                    "Number of tweets : " + $scope.tweetFreq[key].length + "</div>"
                );

 countriesLow.push(marker);
continue;
```

**Comment by djmgit** [comment]
@vibhcool Thanks for the suggestion. I followed it and re factored my code. It helped in reducing code repetition. Please re-review :)

**Comment by vibhcool** [comment]
@djmgit , could you see the scroller, comment if it can be solved?

**Comment by djmgit** [comment]
Test link: http://loklakapps.surge.sh/CountryTweetMap/
@vibhcool please check once again, I have updated my deployment.

**Pull request merged**